### PR TITLE
Add support for DAML-LF's enum types to damlc

### PR DESCRIPTION
**Pull request by hurryabit**
This is part of #105.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/1824)

### CODE REVIEW

**Comment by hurryabit** [comment]
FIY @bitonic @remyhaemmerle-da

File: daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/UtilLF.hs [hunk 1]
```diff
@@ -167,5 +159,22 @@ ghcTypes = Module
           (ERecCon (TypeConApp (qual (dataTypeCon dataProxy)) [TVar (mkTypeVar "a")]) [])
       }
 
+mkVariantOrEnumDataCons :: Version -> [VariantConName] -> DataCons
+mkVariantOrEnumDataCons version cons
```
Comment (ndmitchell) [inline daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/UtilLF.hs]: I'd just call this mkDataEnum and treat it as a smart contstructor that knows about the fallback.

File: daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs [hunk 2]
```diff
@@ -50,6 +50,7 @@ startFromExpr seen world e = case e of
     LF.ERecProj _ _ recEx -> startFromExpr seen world recEx
     LF.ETupleUpd _ recExpr recUpdate -> startFromExpr seen world recExpr `Set.union` startFromExpr seen world recUpdate
     LF.EVariantCon _ _ varg -> startFromExpr seen world varg
```
Comment (ndmitchell) [inline daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs]: Generics required!

File: daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/UtilLF.hs (hunk 1 above)
Comment (hurryabit) [reply daml-foundations/daml-ghc/ghc-compiler/src/DA/Daml/GHC/Compiler/UtilLF.hs]: Make sense.

File: daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs (hunk 2 above)
Comment (hurryabit) [reply daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs]: I guess that's @akalburgi's business. ;)

Comment (ndmitchell) [reply daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs]: Agreed! More a remark for him that the lack of generics is starting to have a cost on other changes.

Comment (akalburgi) [reply daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs]: Currently working on it

Comment (akalburgi) [reply daml-foundations/daml-tools/da-hs-daml-cli/DA/Cli/Visual.hs]: @hurryabit using uniplate PR https://github.com/digital-asset/daml/pull/1826

**Pull request merged**